code/game/objects/items/devices/sound_synth.dm

```diff
@@ -1,79 +1,117 @@
-    volume = text2num(assblast["volume"])
+	set category = "Object"
+	set name = "Select Sound Playback"
+	var/new_sound = tgui_input_list(usr, "Pick a sound!", "Sound Synthesizer", sound_list, default = selected_sound_name)
+	if(!new_sound || !sound_list[new_sound])
+		return
+	to_chat(usr, "<span class='notice'>Sound playback set to: <b>[new_sound]</b>!</span>")
+	selected_sound_name = new_sound
+	var/list/assblast = params2list(sound_list[new_sound])
+	selected_sound = sound(assblast["selected_sound"])
+	shiftpitch = text2num(assblast["shiftpitch"])
+	volume = text2num(assblast["volume"])
+	SSblackbox.record_feedback("tally", "synth_sound_selected", 1, selected_sound_name)
 
-/obj/item/soundsynth/attack_self(mob/user as mob)
-    if(spam_flag + 2 SECONDS < world.timeofday)
-        playsound(src, selected_sound, volume, shiftpitch)
-        spam_flag = world.timeofday
+/obj/item/soundsynth/attack_self(mob/user)
+	if(!selected_sound || !selected_sound_name)
+		to_chat(user, "<span class='warning'>No sound has been selected!</span>")
+		return
+	if(has_cooldown && !COOLDOWN_FINISHED(src, play_cooldown))
+		to_chat(user, "<span class='warning'>You must wait [DisplayTimeText(COOLDOWN_TIMELEFT(src, play_cooldown))] before you can play another sound!</span>")
+		return
+	playsound(user, selected_sound, volume, shiftpitch)
+	SSblackbox.record_feedback("tally", "synth_sound_played", 1, selected_sound_name)
+	if(has_cooldown)
+		COOLDOWN_START(src, play_cooldown, added_cooldown + get_sound_length(user))
 
 /obj/item/soundsynth/AltClick(mob/living/carbon/user)
 	if(!user.canUseTopic(src, BE_CLOSE))
 		return
 	pick_sound()
 
-/obj/item/soundsynth/attack(mob/living/M as mob, mob/living/user as mob, def_zone)
-    if(M == user)
-        pick_sound()
-    else if(spam_flag + 2 SECONDS < world.timeofday)
-        M.playsound_local(get_turf(src), selected_sound, volume, shiftpitch)
-        spam_flag = world.timeofday
-        //to_chat(M, selected_sound) //this doesn't actually go to their chat very much at all.
+/obj/item/soundsynth/attack(mob/living/target, mob/living/user, def_zone)
+	if(target == user)
+		pick_sound()
+		return
+	if(!selected_sound || !selected_sound_name)
+		to_chat(user, "<span class='warning'>No sound has been selected!</span>")
+		return
+	if(has_cooldown && !COOLDOWN_FINISHED(src, play_cooldown))
+		to_chat(user, "<span class='warning'>You must wait [DisplayTimeText(COOLDOWN_TIMELEFT(src, play_cooldown))] before you can play another sound!</span>")
+		return
+	target.playsound_local(get_turf(src), selected_sound, volume, shiftpitch)
+	SSblackbox.record_feedback("tally", "synth_sound_played", 1, selected_sound_name)
+	if(has_cooldown)
+		COOLDOWN_START(src, play_cooldown, added_cooldown + get_sound_length(target))
+
+// WHY BYOND WHYYYYY, WHY DO I HAVE THE QUERY THE CLIENT FOR THE SOUND, CAN'T YOU JUST FILL OUT LEN WHENEVER I MAKE A SOUND DATUM?!
+/obj/item/soundsynth/proc/get_sound_length(mob/living/hearer)
```

Comment:
Cache all of the sounds at once when the first query is made rather than caching only the played sound.

Reply:
You can't, you can only get the length of a sound by _sending it to a client and then querying said client_ - `sound.len` is zero unless `sound` is from `/client/SoundQuery()`

Reply:
tl;dr: byond moment